### PR TITLE
Use stable outputs in illink targets

### PR DESCRIPTION
**Pull request by ViktorHofer**
While building locally multiple times, I noticed that the incremental build of the illink-targets are now to aggressive and could cause the targets to either never run or too often. This could be the case when the linker doesn't trim anything and copies the inputs to the outputs directory with preserving the timestamps. In such cases the inputs would always be newer than the outputs and the step would alway run. Also the path that was used to construct the outputs was missing a trailing directory separator. Using a marker file instead to not be dependent on the freshness of the linker output.

### CODE REVIEW

**Comment by ViktorHofer** [comment]
@safern @ericstj thanks for reviewing the other PRs. Can you please take a look at this one as well?

File: src/libraries/illink-oob.targets [hunk 1]
```diff
@@ -4,7 +4,8 @@
           DependsOnTargets="PrepareForAssembliesTrim">
 
     <PropertyGroup>
-      <OOBAssembliesTrimmedArtifactsPath>$([MSBuild]::NormalizePath('$(ILLinkTrimAssemblyArtifactsRootDir)', 'trimmed-oobs'))</OOBAssembliesTrimmedArtifactsPath>
+      <OOBAssembliesTrimmedArtifactsPath>$([MSBuild]::NormalizeDirectory('$(ILLinkTrimAssemblyArtifactsRootDir)', 'trimmed-oobs'))</OOBAssembliesTrimmedArtifactsPath>
+      <OOBAssembliesMarkerFile>$(ILLinkTrimAssemblyArtifactsRootDir)oob-marker.txt</OOBAssembliesMarkerFile>
```
Comment (akoeplinger) [inline src/libraries/illink-oob.targets:8]: are we sure we want this in the bin and not the obj directory?

File: src/libraries/illink-oob.targets [hunk 2]
```diff
@@ -24,7 +25,7 @@
       <OOBAssemblyToIgnore Include="System.Speech" />
 
       <NetCoreAppCurrentAssembly Include="$(NetCoreAppCurrentRuntimePath)*.dll"
-                                 Exclude="$(NetCoreAppCurrentRuntimePath)*.Generator.dll;
+                                 Exclude="$(NetCoreAppCurrentRuntimePath)*.Generator*.dll;
```
Comment (akoeplinger) [inline src/libraries/illink-oob.targets:28]: why is this needed?

Comment (ViktorHofer) [reply src/libraries/illink-oob.targets:28]: The generator assembly namens aren't consistent in their naming. There was at least one generator that wasn't excluded. I believe it was CoreLib's.

File: src/libraries/illink-oob.targets (hunk 1 above)
Comment (ViktorHofer) [reply src/libraries/illink-oob.targets:8]: Good point. Will move to BaseIntermediateOutputPath.

File: src/libraries/illink-oob.targets (hunk 2 above)
Comment (akoeplinger) [reply src/libraries/illink-oob.targets:28]: ah I see there's `System.Private.CoreLib.Generators.dll` indeed

File: src/libraries/illink-oob.targets (hunk 1 above)
Comment (ViktorHofer) [reply src/libraries/illink-oob.targets]: done

File: src/libraries/illink-oob.targets [hunk 3]
```diff
@@ -4,7 +4,8 @@
           DependsOnTargets="PrepareForAssembliesTrim">
 
     <PropertyGroup>
-      <OOBAssembliesTrimmedArtifactsPath>$([MSBuild]::NormalizePath('$(ILLinkTrimAssemblyArtifactsRootDir)', 'trimmed-oobs'))</OOBAssembliesTrimmedArtifactsPath>
+      <OOBAssembliesTrimmedArtifactsPath>$([MSBuild]::NormalizeDirectory('$(ILLinkTrimAssemblyArtifactsRootDir)', 'trimmed-oobs'))</OOBAssembliesTrimmedArtifactsPath>
+      <OOBAssembliesMarkerFile>$(BaseIntermediateOutputPath)oob-linker-marker.txt</OOBAssembliesMarkerFile>
```
Comment (ericstj) [inline src/libraries/illink-oob.targets:8]: Does this path encode sufficient information such that it is unique for a given set of inputs? EG: configuration, target framework, runtime etc -- not sure all the variables going into the inputs.  Perhaps sharing sample values of the inputs and outputs would help confirm.

Comment (ViktorHofer) [reply src/libraries/illink-oob.targets:8]: It doesn't encode any of the properties you mentioned above. Would it be better to use IntermediateOutputPath instead?
The inputs are dependent on these, you are right: `NetCoreAppCurrentRuntimePath = C:\git\runtime3\artifacts\bin\runtime\net7.0-windows-Debug-x64\`

**Comment by ViktorHofer** [comment]
I addressed the feedback, please review. Thanks

**Comment by ViktorHofer** [comment]
@ericstj @safern can you please review and approve?

I also have another few changes waiting for feedback that I would really love to get in:
- https://github.com/dotnet/runtime/pull/64508
- https://github.com/dotnet/runtime/pull/64500
- https://github.com/dotnet/runtime/pull/64518

As soon as those are in, I can finish implementing https://github.com/dotnet/runtime/pull/64000 and with that probably again spend less time contributing to dotnet/runtime until April ;)

File: src/libraries/illink-oob.targets (hunk 2 above)
Comment (ericstj) [reply src/libraries/illink-oob.targets:28]: I'm surprised we're placing generators in the runtime path, that's wrong.  They don't even target the same thing.  We should see about fixing that (separately is fine).

**Review by ericstj — APPROVED** [review]
